### ISSUE TITLE
Lint when return type `void` is overridden by `Future<T>` for some `T`?

### ISSUE THREAD

**Issue by eernstg** [dart-lang/linter#1882]
Cf. https://github.com/dart-lang/language/issues/735.

It is error-prone and not likely to be useful when the return type `void` is overridden by the return type `Future<T>` or `FutureOr<T>`, for any `T`. For example:

```dart
abstract class Bar {
  void test();
}

class Baz extends Bar {
  Future<void> test() async {
    return Future.delayed(Duration(seconds: 3), () => print('Baz'));
  }
}
```

A lint on such overrides could supplement the support for not accidentally discarding futures that we already have with `unawaited_futures`.

**Comment by bwilkerson** [comment]
@MichaelRFairhurst @jcollins-g (Because this came up in a CL this morning.)

**Comment by bwilkerson** [comment]
I have to ask whether this should be extended to prohibit any override of a method with a return type of `void` with a non-`void` return type. It seems similarly error-prone to ignore any returned value from an overriding method, and of marginal utility to allow an overriding method to return a value (and the latter case could always be handled with an `ignore` comment as a way of documenting that the behavior is intentional).

(That seems especially true for overrides that are marked with the `useResult` annotation, but that might be better addressed by expanding the checking associated with the annotation.)

**Comment by eernstg** [comment]
The argument that usually comes up in favor of allowing a method to have return type `void`, and then be overridden by a method with some other return type, is that the overridden method may not have anything meaningful to return.

A particularly tricky case occurs when the returned result should be ignored because it cannot be interpreted correctly at the more abstract level.

Here's an extreme example where the interpretations are nearly opposite. We would probably never see anything this blatant in real life, but we could still have cases where two different overrides of a method return a result that should be interpreted _somewhat_ differently, and hence it could be useful to use the return type `void` in the case where it isn't known which interpretation is appropriate:

```dart
abstract class Action {
  // We could make it `bool doit();`, but we wouldn't know what that bool means.
  void doit();
}

class ActionThatShouldSucceed implements Action {
  bool doit() {
    ... // Do the thing.
    ... // Return true if everything was done successfully.
  }
}

class ActionThatShouldFail implements Action {
  bool doit() {
    ... // Attempt to do the thing.
    ... // Return true if various checks detected this attempt and stopped it.
  }
}
```

You could of course say that it is just bad design to consider those two `doit` methods to be "the same method", so they should have had different names (and maybe we shouldn't have an `Action` class at all). However, if you want to perform a system stress test you might not care whether it's one or the other kind of action.

One possible conclusion could be that this argument is silly :grin:, and that situation never occurs in real life anyway, and then we should lint _all_ overrides where return type `void` is overridden by any non-void type. Another conclusion could be that the `Future` case and the general case should be handled by two distinct lints.

In any case, I do think it makes sense to lint the situation where a `void` return type is overridden by `Future<T>` or `FutureOr<T>`, because this case has a direct connection to unawaited futures, and we all know that they give rise to subtle bugs.

**Comment by bwilkerson** [comment]
Actually, I'll offer a third conclusion, the one I was alluding to earlier, which is that this code pattern is uncommon enough that it should be documented whenever used and hence it isn't a problem to require part of that documentation to be an `ignore` comment.

I think we should generalize the lint to catch all the cases.

**Comment by eernstg** [comment]
Sounds good!

**Comment by srawlins** [comment]
I need a few more 👍 to get this into the linter. @lrhn @munificent @goderbauer, want to chime in?

**Comment by srawlins** [comment]
Good ideas, @lrhn . It sounds like this might be a "mixed Future and void return type overrides" lint rule. Given a class `B` with a method `m` whose "direct override" is `A.m` (`B.m` directly overrides `A.m` via extension, implementation, or mixing-in) we would report a lint when:

* the return type of `A.m` is exactly `void`, and the return type of `B.m` is not exactly `void`,
* the return type of `A.m` is `Future<void>`and the return type of `B.m` is `Future<T>` for some `T` which is not `void` (another criterion added by @eernstg in a PR) (how to handle subtypes of Future, like CancelableFuture? Analyzer has concepts like "as instance of" but I don't know how this maps to type theory. We'd use something like "if the return type of `A.m` is a subtype of `Future<void>` and _as_ a `Future<T>` yields some `T` which is `void`, _and_ the return type of `B.m` is a subtype of `Future<void>` and _as_ a `Future<T>` yields some `T` which is _not_ `void`, ... and I imagine we have to flatten as well)
* the return type of `A.m` is not `Future<T>` or `FutureOr<T>` for any `T`, and the return type of `B.m` is `Future<T>` or `FutureOr<T>` for some `T` (how to handle subtypes of Future, like CancelableFuture? Luckily FutureOr cannot be subtyped)

**Comment by lrhn** [comment]
> we would report a lint when:

> * the return type of `A.m` is exactly `void`, and the return type of `B.m` is not exactly `void`,

I wouldn't warn about that. That's very much intended to work, and the reason `void` is a supertype of other types to begin with.
What you should not have is a `B` being marked as `@useResult` or returning a `Future` or `FutureOr` type (which can implicitly be thought of as must-be-used types, at least if the `unawaited_futures` lint is enabled).

One could argue that having a superclass method without `@useResult` must not have a subclass method with `@useResult`, because casting to the superclass would suggest that the result doesn't need to be used.
Maybe that's too strict, but ... probably not.

> * the return type of `A.m` is `Future<void>`and the return type of `B.m` is `Future<T>` for some `T` which is not `void`  

Wouldn't warn there either. That's just the asynchronous variant of the above.
The future itself should be handled, after that, it's OK to 

Does `@useResult` on an asynchronous function apply to the future or to the future's value? I think it *should* be the latter, because that's actually useful and the former is handled by `unawaited_futures` instead.

> * the return type of `A.m` is not `Future<T>` or `FutureOr<T>` for any `T`, and the return type of `B.m` is `Future<T>` or `FutureOr<T>` for some `T` (how to handle subtypes of Future, like CancelableFuture? Luckily FutureOr cannot be subtyped)

This one I'd subscribe to.
I'd treat any subtype of `Future` as a future. Say a type is futurish if:
* Type is `FutureOr<T>` for some future.
* Type implements `Future<T>` for some `T`.
* Type is `S?` and `S` is futurish.

Basically, we consider a type to be "probably a future" if it directly implements `Future`, or is a union types where one of the types does. _Should we include sealed types where one of the immediate subtypes implements `Future`? Should we include inline classes with representation types which are futures? (Probably no to both, but worth considering.)_

Then we can warn if the return type of `B.m` is futurish and the return type of `A.m` is not.

**Comment by srawlins** [comment]
It sounds like this warrants a lot of discussion then. I won't be working on an implementation any time soon.

**Comment by eernstg** [comment]
> Say a type is futurish if: ...

We could use [the following concept](https://github.com/dart-lang/language/blob/a374667bcc4095edd22e2e6d454cbe9d9f9d85fa/specification/dartLangSpec.tex#L11309) from the language specification to define which types are 'futurish':

<img width="649" alt="Screenshot 2023-02-25 at 12 37 17" src="https://user-images.githubusercontent.com/9606994/221625691-55403b4d-c9a8-4647-8e06-5f3d7f6adab5.png">

So here is one possible interpretation of the desired semantics:

> Lint a member if it has a return type which has a future type, and it directly or indirectly overrides a member whose return type is `void`.

This is very nearly my original proposal in this issue. Another proposal which came up was:

> Lint a member if it has a return type which is not `void`, and it overrides a member whose return type is `void`.

@lrhn, you don't want to lint this situation. I tried to come up with a description of some situations where this particular kind of overriding is useful, [here](https://github.com/dart-lang/linter/issues/1882#issuecomment-1427558227). However, it does get somewhat awkward. ;-)

Do you have at least one actually useful example where `void` is overridden by non-`void`? Or should we just use the meta-rule "someone might come up with a good usage" and then abstain from linting it simply because it might happen?

The last rule which was mentioned is actually a variant of the second rule, linting at the return type `Future<T>` when it overrides `Future<S>` and rule 2 would lint `T` overridden by `S`. ("Same object, different timing".)

We can of course have a restricted version of this latest rule such that it puts the focus on futures:

> Lint a member if it has a return type which has a future type with a type argument that has a future type, and it overrides a member with a return type that has a future type with the type argument `void`.

**Comment by lrhn** [comment]
I don't want to warn about overriding `void` return with another type of return, as long as the other type is not
* a future type (and maybe the `unawaited_futures` lint is enabled, but isn't it always?), or
* marked `@useResult` (but that may apply to any supermethod not marked `@useResult`, not just `void`-returning ones.)

I don't have a concrete example. I'm sure we have some in the platform libraries. Also some where we *did* override to return a future in a subclass. That's where we learned that it was a bad idea.
The rest are fine.

it's not something that happens often, because returning a value that is so easily ignorable suggests that the value isn't important, so why are you returning it to being with? We don't do traditional function chaining in Dart, and if it makes logical sense for a function to return a value (say the old or new value for an update operation like `Map.putIfAbsent`), it usually also makes sense for the superclass.

Still, there is no *danger* in overriding a method with a more precise type. We don't warn if a superclass returns `Object` and a subclass returns `int`, because it's safe. Anyone using a subclass instance at the superclass type can definitely handle an `int`, since they must be ready to handle` any `Object`.
Likewise, there is no danger in overriding a `void` returning function with an `int` returning function. Anyone using a subclass instance at the superclass type can handle an `int` being returned, because they can handle any `Object?` being returned ... by ignoring it. 
Ignoring a value is all fine and dandy, unless it's a future or a `@useResult` value.

Adding a warning in that case has _no benefit_. It's not protecting anyone from anything. I can see no reason to enable such a lint.

A lint should either:
* Protect you against a certain or likely error-pattern (like `unawaited_futures` and the "unrelated types" lints). It's not invalid code, but it's either probably wrong, or the severity if wrong is high enough, so we want you to explicitly state that you know what you're doing (`unawaited`, `// ignore`).
* Guide you towards a preferred alternative (style lints, suggesting more specific operations like "use `isEmpty` or use collection literals).
* Enforce a specific style choice (like previous point, but locally preferred instead of globally, like single quotes or final locals).

There should be either few false positives, or the severity of a mistake is so great that it's worth extra effort to avoid it.

In every case, there should be a workaround that allows you to continue without the lint warning. Either rewrite to follow the preferred style, fix the actual problem, or acknowledge that you're using a high-risk operation deliberately.

I don't see which category the "subclass returning a non-`void` type where the superclass doesn't" issue falls into.

And I don't see a simple workaround. 
Sure, you can just not return the value, but I assume you do it for a reason.
And you might not be able to change the superclass to return the same value.
(I guess there is always `// ignore:`, but you should be able to document the reason the `ignore` is safe here, and the documentation here really is just "I want to return a value".)


So yes, I don't want to lint that situation. :wink:

**Comment by eernstg** [comment]
OK, but then I'd like to recommend that we go back to, essentially, the original proposal: Lint cases where an instance member declaration with return type `void` is overridden by a declaration whose return type has a future type.

The definition of 'has a future type' can be found [here](https://github.com/dart-lang/linter/issues/1882#issuecomment-1446702515).

**Comment by eernstg** [comment]
You do have a point there!

However, if we're going to flag `Object` being overridden by `Future<int>` then we should perhaps also flag `Object o = myFuture;` in `unawaited_futures`, and even `MyTypeVariable x = myFuture`?

**Comment by lrhn** [comment]
We (probably) don't warn for `Object o = myFuture;` because it's locally visible that you are doing it. The expression of type `Future` is there, so we expect you to know it's a future, and you are storing it in a variable for later, so you probably intend to do something with it then. (Internally, in the `_Future` implementation, I probably store futures in reused `Object?` variables, along with a state saying that it contains a future, so I can get it out and cast it back later.)

For subclass overrides, a `foo(SuperClass o) { o.voidMethod(); }` won't have a chance to see that `SubClass.voidMethod` returns a future. That's why we have to warn the `SubClass.voidMethod` declaration that it's likely to get its returned future ignored.

We could warn in the `Object o = myFuture;` case too, but there is no workaround, like using `unawaited`, so you'd have to use `// ignore`. Or `var o = myFuture as Object;`.
Probably not worth it.

**Comment by eernstg** [comment]
> Probably not worth it.

I think this conclusion delineates and preserves a specific inconsistency in our treatment of this topic: We aren't really trying to catch situations where a future is "forgotten" by being assigned to a variable/parameter whose type doesn't have a future type. For example:

```dart
abstract class A { Object foo(); }
class B implements A { Future<int> foo() {...}} // LINT?

void bar(Object o) {}

Future<int> get e => Future.value(1);

void main() async {
  Object o1 = e; // Silent. We can see the type `Object`, but not the type of `e`.
  bar(e); // Silent. No visible hint that we're forgetting that `e` is a future.

  A a = B(); // Very indirect: We're forgetting that `B.foo` returns a future.
  var o2 = a.foo(); // Every call forgets a future.

  void ignore<X>(X x) {}
  ignore(e); // We're forgetting a future again.

  List<X> preserve<X>(X x) => [x]; // Or `[]`.
  var e2 = preserve(e); // We didn't forget the future. (Or maybe we did?)
}
```

I'm tempted to recommend that we address the broader issue "upcast forgets future-ness" separately, and then focus on `void` and futures in this lint.

A future "upcast forgets future-ness" lint could then target all those situations where the future-ness of a given expression is unambiguously forgotten. That is, we'd probably ignore cases like `preserve`, but we would flag `Object o = e;` and such, because there is no doubt that the future-ness of `e` is forgotten when that object is accessed using `o`. The case where a without-future return type is overridden by a with-future return type would be covered by this lint, too.

The "upcast forgets future-ness" lint could be somewhat noisy (that is, it might have many false positives), but that would not be so much of a problem if it's a separate lint (which is not included in 'recommended' and such).

**Comment by a-siva** [comment]
Is https://github.com/flutter/flutter/issues/120112 somewhat related to the issue being discussed here ? I was wondering if the lint/error being suggested there makes sense.

**Comment by eernstg** [comment]
@a-siva wrote:
> somewhat related

Very much the same thing, thanks! I added a comment there.
